### PR TITLE
Auto-export GDML from Geant4 geometry

### PR DESCRIPTION
**Pull request by sethrj**
This is a stopgap measure before #557 . It adds a helper class to export the Geant4 geometry from a "world" volume as a GDML, and adapts the interface from #557 along with the unit test harness updates I made. The new test geometry `solids.gdml` (from Geant4) is chock full of errors, though, so it's not a very useful test except to show that the results will change when Guilherme's work is finished.

Because G4's GDML exporter simply exits if it's not the master thread, we have to ensure that the `SharedParams` initialization happens on the master thread, so that logic has changed.

Furthermore, to allow us to test the integration on HIP and/or with ORANGE, I've added a superficial G4 constructor and added some hacky code to propagate the GDML file into the `accel` helper functions.

---

VecGeom failures on `solids.gdml`:
```
Middleware::processNode: an unknown solid genericPolycone with name testgenpoly
Middleware::processNode: an unknown solid genericPolyhedra with name testgenph
Middleware::processNode: an unknown solid twistedbox with name ttwistedbox
Middleware::processNode: an unknown solid twistedtrd with name ttwistedtrd
Middleware::processNode: an unknown solid twistedtrap with name ttwistedtrap
Middleware::processNode: an unknown solid twistedtubs with name ttwistedtubs
processLogicVolume: Could not find solid reference testgenpoly for logical volume vol412
processLogicVolume: Could not find solid reference ttwistedbox for logical volume vol71
processLogicVolume: Could not find solid reference ttwistedtrd for logical volume vol81
processLogicVolume: Could not find solid reference ttwistedtrap for logical volume vol91
processLogicVolume: Could not find solid reference ttwistedtubs for logical volume vol101
processLogicVolume: Could not find solid reference testgenph for logical volume vol502
Middleware::processPhysicalVolume: could not find volume vol412
```

And since the "real" GDML has very weird traces that the "read through Geant4 and exported" doesn't, I think there's also additional bugs in the VGDML reading.

See [VecGeom issue 427](https://sft.its.cern.ch/jira/browse/VECGEOM-427) (CERN users only)

### CODE REVIEW

**Review by mrguilima — APPROVED** [review]
All looks good.
Thanks, Seth!

**Review by mrguilima — APPROVED** [review]
Sorry.. clicked on some review reset button, had to re-review it. :-(

**Review by amandalund — COMMENTED** [review]
Ditto, changes look great @sethrj! However, the new VecGeom tests are crashing for me (using VecGeom version 1.1.17) when loading the geometry from solids.gdml:
```
[----------] 1 test from SolidsTest
[ RUN      ] SolidsTest.accessors
info: Loading VecGeom geometry from GDML at /home/alund/celeritas_project/celeritas/test/celeritas/data/solids.gdml

 *** Break *** segmentation violation
 Generating stack trace...
 0x00007fec4429065e in vgdml::Middleware::processNode(xercesc_3_2::DOMNode const*) at /tmp/alund/spack-stage/spack-stage-vecgeom-1.1.17-pkgoej62uf42z3xtn3rwhep3nvekqzwn/spack-src/persistency/gdml/source/src/Middleware.cpp:195 from /home/alund/spack/var/spack/environments/celeritas/.spack-env/view/lib64/libvgdml.so
 0x00007fec4429056d in vgdml::Middleware::processNode(xercesc_3_2::DOMNode const*) at /tmp/alund/spack-stage/spack-stage-vecgeom-1.1.17-pkgoej62uf42z3xtn3rwhep3nvekqzwn/spack-src/persistency/gdml/source/src/Middleware.cpp:208 from /home/alund/spack/var/spack/environments/celeritas/.spack-env/view/lib64/libvgdml.so
 0x00007fec4429056d in vgdml::Middleware::processNode(xercesc_3_2::DOMNode const*) at /tmp/alund/spack-stage/spack-stage-vecgeom-1.1.17-pkgoej62uf42z3xtn3rwhep3nvekqzwn/spack-src/persistency/gdml/source/src/Middleware.cpp:208 from /home/alund/spack/var/spack/environments/celeritas/.spack-env/view/lib64/libvgdml.so
 0x00007fec44282c8b in vgdml::Parser::Load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, double, bool) at /tmp/alund/spack-stage/spack-stage-vecgeom-1.1.17-pkgoej62uf42z3xtn3rwhep3nvekqzwn/spack-src/persistency/gdml/source/src/Frontend.cpp:36 from /home/alund/spack/var/spack/environments/celeritas/.spack-env/view/lib64/libvgdml.so
 0x00007fec44282ee9 in vgdml::Frontend::Load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, double, bool) at /soft/compilers/gcc/8.2.0/linux-rhel7-x86_64/include/c++/8.2.0/bits/unique_ptr.h:150 from /home/alund/spack/var/spack/environments/celeritas/.spack-env/view/lib64/libvgdml.so
 0x00007fec4758f37e in celeritas::VecgeomParams::VecgeomParams(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) at /home/alund/celeritas_project/celeritas/src/celeritas/ext/VecgeomParams.cc:50 from /home/alund/celeritas_project/celeritas/build-rev/lib64/libceleritas.so
...
```

**Comment by sethrj** [comment]
thanks Amanda! I'll put a skip on that one conditional on newer versions.

File: test/celeritas/ext/Vecgeom.test.cc [hunk 1]
```diff
@@ -425,35 +413,157 @@ TEST_F(FourLevelsTest, TEST_IF_CELERITAS_CUDA(device))
...
+#define SolidsGeantTest TEST_IF_CELERITAS_GEANT(SolidsGeantTest)
+class SolidsGeantTest : public GeantBuilderTestBase
+{
+  public:
+    const char* geometry_basename() const final { return "solids"; }
+};
+
+//---------------------------------------------------------------------------//
+
+TEST_F(SolidsGeantTest, accessors)
```
Comment (amandalund) [inline test/celeritas/ext/Vecgeom.test.cc:510]: Need a skip for this one too 😉

Comment (sethrj) [reply test/celeritas/ext/Vecgeom.test.cc:510]: Boo I had hoped it would be a bug in the importer for the original, not the re exported one. :(

Comment (amandalund) [reply test/celeritas/ext/Vecgeom.test.cc:510]: Yeah, unfortunately it crashes loading the Geant4-exported one too.

**Pull request merged**